File: src/main/java/com/joecollins/graphics/components/ListingFrameBuilder.java
package com.joecollins.graphics.components;

import com.joecollins.bindings.BindableList;
import com.joecollins.bindings.Binding;
import com.joecollins.bindings.IndexedBinding;
import java.awt.Color;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class ListingFrameBuilder {

  private BarFrame barFrame = new BarFrame();

  private final List<Binding<?>> bindings = new ArrayList<>();

  public static <T> ListingFrameBuilder of(
      BindableList<T> list,
      Function<T, String> leftTextFunc,
      Function<T, String> rightTextFunc,
      Function<T, Color> colorFunc) {
    ListingFrameBuilder builder = new ListingFrameBuilder();
    BarFrame barFrame = builder.barFrame;
    barFrame.setNumBarsBinding(Binding.sizeBinding(list));
    barFrame.setLeftTextBinding(IndexedBinding.propertyBinding(list, leftTextFunc::apply));
    barFrame.setRightTextBinding(IndexedBinding.propertyBinding(list, rightTextFunc::apply));
    barFrame.addSeriesBinding(
        "Item",
        IndexedBinding.propertyBinding(list, colorFunc::apply),
        IndexedBinding.propertyBinding(list, x -> 1));
    return builder;
  }

  public static <T> ListingFrameBuilder of(
      Binding<List<T>> list,
      Function<T, String> leftTextFunc,
      Function<T, String> rightTextFunc,
      Function<T, Color> colorFunc) {
    BindableList<T> bindableList = new BindableList<>();
    list.bindLegacy(bindableList::setAll);
    ListingFrameBuilder ret = of(bindableList, leftTextFunc, rightTextFunc, colorFunc);
    ret.bindings.add(list);
    return ret;
  }

  public static <T> ListingFrameBuilder ofFixedList(
      List<T> list,
      Function<T, Binding<String>> leftTextFunc,
      Function<T, Binding<String>> rightTextFunc,
      Function<T, Binding<Color>> colorFunc) {
    ListingFrameBuilder builder = new ListingFrameBuilder();
    BarFrame barFrame = builder.barFrame;
    barFrame.setNumBarsBinding(Binding.fixedBinding(list.size()));
    barFrame.setLeftTextBinding(IndexedBinding.listBinding(list, leftTextFunc::apply));
    barFrame.setRightTextBinding(IndexedBinding.listBinding(list, rightTextFunc::apply));
    barFrame.addSeriesBinding(
        "Item",
        IndexedBinding.listBinding(list, colorFunc::apply),
        IndexedBinding.listBinding(list, x -> Binding.fixedBinding(1)));
    return builder;
  }

  public ListingFrameBuilder withHeader(Binding<String> headerBinding) {
    barFrame.setHeaderBinding(headerBinding);
    return this;
  }

  public ListingFrameBuilder withSubhead(Binding<String> subheadBinding) {
    barFrame.setSubheadTextBinding(subheadBinding);
    return this;
  }

  public ListingFrameBuilder withNotes(Binding<String> notesBinding) {
    barFrame.setNotesBinding(notesBinding);
    return this;
  }

  public BarFrame build() {
    return barFrame;
  }
}

File: src/main/java/com/joecollins/graphics/components/ListingFrameBuilder.kt
package com.joecollins.graphics.components

import com.joecollins.pubsub.asOneTimePublisher
import com.joecollins.pubsub.combine
import com.joecollins.pubsub.map
import com.joecollins.pubsub.merge
import java.awt.Color
import java.util.concurrent.Flow

class ListingFrameBuilder {

    private var headerPublisher: Flow.Publisher<out String?>? = null
    private var subheadTextPublisher: Flow.Publisher<out String?>? = null
    private var notesPublisher: Flow.Publisher<out String?>? = null
    private var barsPublisher: Flow.Publisher<out List<BarFrame.Bar>>? = null

    fun withHeader(headerPublisher: Flow.Publisher<out String?>): ListingFrameBuilder {
        this.headerPublisher = headerPublisher
        return this
    }

    fun withSubhead(subheadPublisher: Flow.Publisher<out String?>): ListingFrameBuilder {
        this.subheadTextPublisher = subheadPublisher
        return this
    }

    fun withNotes(notesPublisher: Flow.Publisher<out String?>): ListingFrameBuilder {
        this.notesPublisher = notesPublisher
        return this
    }

    fun build(): BarFrame {
        return BarFrame(
            headerPublisher = headerPublisher ?: (null as String?).asOneTimePublisher(),
            subheadTextPublisher = subheadTextPublisher,
            notesPublisher = notesPublisher,
            barsPublisher = barsPublisher ?: emptyList<BarFrame.Bar>().asOneTimePublisher()
        )
    }

    companion object {
        @JvmStatic
        fun <T> of(
            list: Flow.Publisher<out List<T>>,
            leftTextFunc: (T) -> String,
            rightTextFunc: (T) -> String,
            colorFunc: (T) -> Color
        ): ListingFrameBuilder {
            val builder = ListingFrameBuilder()
            builder.barsPublisher = list.map { l ->
                l.map {
                    BarFrame.Bar(leftTextFunc(it), rightTextFunc(it), null, listOf(Pair(colorFunc(it), 1)))
                }
            }
            return builder
        }

        @JvmStatic fun <T> ofFixedList(
            list: List<T>,
            leftTextFunc: (T) -> Flow.Publisher<out String>,
            rightTextFunc: (T) -> Flow.Publisher<out String>,
            colorFunc: (T) -> Flow.Publisher<out Color>
        ): ListingFrameBuilder {
            val builder = ListingFrameBuilder()
            builder.barsPublisher =
                list.map {
                    leftTextFunc(it).merge(rightTextFunc(it)) { left, right -> Pair(left, right) }
                        .merge(colorFunc(it)) {
                            (left, right), color ->
                            BarFrame.Bar(left, right, listOf(Pair(color, 1)))
                        }
                }
                    .combine()
            return builder
        }
    }
}
